_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,282 @@
+
+## Contravariant
+Our `Functor` interface made explicit the behavior of covariance - we can define a similar interface that
+captures contravariant behavior. Remember that contravariance is covariance with the direction flipped,
+which translates to the direction of the function being flipped.
+
+```scala
+trait Contravariant[F[_]] {
+  def contramap[A, B](fa: F[A])(f: B => A): F[B]
+}
+```
+
+We can implement an instance for `Show`.
+
+```scala
+val showContravariant: Contravariant[Show] =
+  new Contravariant[Show] {
+    def contramap[A, B](fa: Show[A])(f: B => A): Show[B] =
+      new Show[B] {
+        def show(b: B): String = {
+          val a = f(b)
+          fa.show(toShow)
+        }
+      }
+  }
+```
+
+Here we are saying if we can show an `A`, we can show a `B` by turning a `B` into an `A` before showing it.
+Upcasting is a specific case of this, when `B` is a subtype of `A`.
+
+```scala
+def contraUpcast[F[_], A, B >: A](contra: Contravariant[F], fb: F[B]): F[A] =
+  contra.contramap(fb)((a: A) => a: B)
+```
+
+Going back to `Shape`s and `Circle`s, we can show a `Circle` by upcasting it into a `Shape` and showing that.
+
+# Invariance
+Unannotated type parameters are considered invariant - different instantiations of a type constructor have no
+relationship with one another regardless of what relationships the type parameters may have. Given invariant
+`F[_]`, an `F[Circle]` is not a subtype of `F[Shape]` - you need to explicitly provide the conversion.
+
+## Array once more
+`Array`s are invariant in Scala because they can be neither covariant nor contravariant. If we make it
+covariant, we can get unsafe writes. If we make it contravariant, we can get unsafe reads. Since
+read-only types can only be covariant and write-only types contravariant, our compromise is to make
+types that support both invariant.
+
+Therefore in order to treat an `Array` of one type as an `Array` of another, we need to have a way of converting
+in both directions. This must be provided manually as the type checker has no way of knowing what the
+conversion would be.
+
+## Invariant
+Similar to (covariant) `Functor` and `Contravariant`, we can write `Invariant`.
+
+```scala
+trait Invariant[F[_]] {
+  def imap[A, B](fa: F[A])(f: A => B)(g: B => A): F[B]
+}
+```
+
+For demonstration purposes we write our own `Array` type
+
+```scala
+class Array[A] {
+  private var repr = ListBuffer.empty[A]
+
+  def read(i: Int): A =
+    repr(i)
+  def write(i: Int, a: A): Unit =
+    repr(i) = a
+}
+```
+
+and define `Invariant[Array]`.
+
+```scala
+val arrayInvariant: Invariant[Array] =
+  new Invariant[Array] {
+    def imap[A, B](fa: Array[A])(f: A => B)(g: B => A): Array[B] =
+      new Array[B] {
+        // Convert read A to B before returning - covariance
+        override def read(i: Int): B =
+          f(fa.read(i))
+
+        // Convert B to A before writing - contravariance
+        override def write(i: Int, b: B): Unit =
+          fa.write(i, g(a))
```

Comment:
Here I was trying to explain why `Array` must be invariant since I was talking about how it couldn't be co- or contra-variant early on. Good call on adding an example that doesn't require mutation though, maybe to keep it consistent I'll write a `Serializer` type class that combines both `Read` and `Show`:

```scala
trait Serializer[A] extends Read[A] with Show[A]
```